### PR TITLE
(DOCSP-16245): [iOS] SwiftUI AsyncOpen and AutoOpen docs

### PR DESCRIPTION
**Pull request by dacharyc**
### Jira

- https://jira.mongodb.org/browse/DOCSP-16245

### Staged Changes (Requires MongoDB Corp SSO)

- [Sync Changes Between Devices](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-16245/sdk/ios/examples/sync-changes-between-devices/#open-a-synced-realm-with-swiftui): New section: "Open a Synced Realm with SwiftUI"

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 1]
```diff
@@ -328,6 +328,188 @@ users from opening a realm while offline.
...
 
+.. _ios-open-synced-realm-swiftui:
+
+Open a Synced Realm with SwiftUI
+--------------------------------
+
+.. versionadded:: 10.12.0
+
+SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
+populate views. These property wrappers provide a syntax that integrates nicely
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:339]: suggestion: remove editorial flavor
```suggestion
populate views. These property wrappers provide a syntax that integrates
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 2]
```diff
@@ -328,6 +328,188 @@ users from opening a realm while offline.
...
+
+SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
+populate views. These property wrappers provide a syntax that integrates nicely 
+with SwiftUI to let developers declare whether their client app should:
+
+- Only open a {+realm+} if they can download updates from the server first. 
+  This requires a successful network connection.
+- Attempt to download updates from the server, but open a {+realm+} offline if 
+  there is no network connection. This supports an offline-first approach at
+  the cost of potentially working with stale data.
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:346]: Note: This reads more like a flowchart at the moment than describing the provided wrappers. As is, the flow I interpret is
```
    if exists(networkConnection) {
      updateFromRemote()
    } else {
      useLocalRealm()
    }
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 3]
```diff
@@ -328,6 +328,188 @@ users from opening a realm while offline.
...
+
+{+app+}s that need to always have up-to-date information from the server, such
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:348]: suggestion: simplify

```suggestion
{+app+}s that require up-to-date information from the server, such
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 4]
```diff
@@ -328,6 +328,188 @@ users from opening a realm while offline.
...
+as game apps with live leaderboards that the user can play on multiple devices, 
+should use ``@AsyncOpen`` to open a {+realm+}. This only works if the user has 
+a network connection when opening the {+realm+}, but ensures the user is never 
+using the app with stale data.
+
+{+app+}s that want to enable offline-first capability, where it's not a problem
+for the user to work with potentially stale data, should use ``@AutoOpen`` to
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:355]: note: the phrasing for this is different than the "online-first" strategy presented above. Maybe reword to not use jargon initially (offline-first), an then use jargon.

> Apps that are tolerant of potentially stale data ...

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 5]
```diff
@@ -328,6 +328,188 @@ users from opening a realm while offline.
...
+
+- Only open a {+realm+} if they can download updates from the server first. 
+  This requires a successful network connection.
+- Attempt to download updates from the server, but open a {+realm+} offline if 
+  there is no network connection. This supports an offline-first approach at
+  the cost of potentially working with stale data.
+
+{+app+}s that need to always have up-to-date information from the server, such
+as game apps with live leaderboards that the user can play on multiple devices, 
+should use ``@AsyncOpen`` to open a {+realm+}. This only works if the user has
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:350]: question: would linking AsyncOpen to the header below be useful here for users that identify this as their desired use-case?

File: source/sdk/ios/examples/sync-changes-between-devices.txt (hunk 4 above)
Comment (terakilobyte) [reply source/sdk/ios/examples/sync-changes-between-devices.txt:355]: same question as the other method, should there be a link here for users to jump directly to their desired use-case?

**Review by terakilobyte — CHANGES_REQUESTED** [review]
A few nits/comments, questions, and suggestions.

File: source/sdk/ios/examples/sync-changes-between-devices.txt (hunk 2 above)
Comment (dacharyc) [reply source/sdk/ios/examples/sync-changes-between-devices.txt:346]: Fair point. I omitted the bullets entirely and rewrote this intro section to be clearer, and more consistent with language used elsewhere on the page.

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 6]
```diff
@@ -328,6 +328,189 @@ users from opening a realm while offline.
...
+
+{+app+}s that require up-to-date information from the server, such as game apps 
+with live leaderboards that the user can play on multiple devices, should 
+:ref:`use @AsyncOpen to open a {+realm+}<swiftui-async-open-synced-realm>`. 
+This only works if the user has a network connection when opening the {+realm+}, 
+but ensures the user is never using the app with stale data.
+
+{+app+}s where it's not a problem for the user to work with potentially stale 
+data, and the developer instead wants to enable offline-first capability,
+should :ref:`use @AutoOpen to open the {+realm+}<swiftui-auto-open-synced-realm>`.
```
Comment (terakilobyte) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:352]: suggestion: This still feels a bit odd/unwieldy. I think it's because there are two clauses here but the previous use-case only has one. Maybe further reword this to introduce it like you did the previous use-case?

**Review by terakilobyte — APPROVED** [review]
LGTM. One suggestion, but it's probably more personal preference than anything and isn't a blocker to landing this.

**Comment by dianaafanador3** [comment]
I think we should mention our new enviromnent value partitionValue
 AsyncOpenPartitionView()
                        .environment(\.partitionValue, <partitionValue>)
which can be used to dynamically inject a partitionValue into @AsyncOpen and @AutoOpen, for example in cases where this value is retrieved from the user id
`AsyncOpenPartitionView()
            .environment(\.partitionValue, user.id)`,
where we get the user id only after loggin in

File: source/sdk/ios/examples/sync-changes-between-devices.txt (hunk 6 above)
Comment (dacharyc) [reply source/sdk/ios/examples/sync-changes-between-devices.txt]: Hmm, fair point. I've updated it again based on this feedback. Thank you!

**Comment by dacharyc** [comment]
Hey, @dianaafanador3 , thank you for pointing this out!

> I think we should mention our new enviromnent value partitionValue

I had missed this completely, but have now added it to the documentation here.

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 7]
```diff
@@ -328,6 +328,230 @@ users from opening a realm while offline.
...
+changesets before opening {+realm+}s, or open {+realm+}s with data on the 
+devices.
+
+{+app+}s that require up-to-date information from the server, such as game apps 
+with live leaderboards that the user can play on multiple devices, should 
+:ref:`use @AsyncOpen to open a {+realm+}<swiftui-async-open-synced-realm>`. 
+This only works if the user has a network connection when opening the {+realm+}, 
+but ensures the user is never using the app with stale data.
+
+{+app+}s where it's not a problem for the user to work with potentially stale
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:350]: I don't think this is the correct use of {+app+}? Aren't you referring to apps in general?
```suggestion
Apps where it's not a problem for the user to work with potentially stale 
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 8]
```diff
@@ -328,6 +328,230 @@ users from opening a realm while offline.
...
+From the LoginView, you can inject the ``.partitionvalue`` environment value 
+into the view where you are performing the ``@AsyncOpen`` to populate the 
+``user.id``:
+
+.. code-block:: swift
+
+   struct LoginView: View {
+
+      var body: some View {
+         ...view code...
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:388]: It's always nice to try to keep it at least syntactically correct:
```suggestion
         // ...view code...
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 9]
```diff
@@ -328,6 +328,230 @@ users from opening a realm while offline.
...
+.. code-block:: swift
+
+   struct AsyncOpenView: View {
+
+      var body: some View {
+         ...view code...
+         ListView()
+            .environment(\.realm, realm)
+      }
+   }
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:407]: This block doesn't look right... where's the @AsyncOpen?

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 10]
```diff
@@ -328,6 +328,230 @@ users from opening a realm while offline.
...
+for the current user. If there is no internet connection, this property wrapper
+instead returns an opened realm for the given ``appId`` and ``partitionValue``.
+
+The property wrapper publishes states, represented by the :swift-sdk:`AsyncOpenState 
+enum <Enums/AsyncOpenState.html#/s:10RealmSwift14AsyncOpenStateO5erroryACs5Error_pcACmF>`,
+which you can use to update the view. 
+
+Pass Environment Values
+```````````````````````
+From the LoginView, you can inject the ``.partitionvalue`` environment value
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:478]:
```suggestion
From the LoginView, you can inject the ``.partitionValue`` environment value 
```

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 11]
```diff
@@ -328,6 +328,230 @@ users from opening a realm while offline.
...
+   @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen
+
+This SwiftUI property wrapper initiates ``Realm.asyncOpen()`` for the current 
+user. The property wrapper publishes states, represented by the :swift-sdk:`AsyncOpenState 
+enum <Enums/AsyncOpenState.html#/s:10RealmSwift14AsyncOpenStateO5erroryACs5Error_pcACmF>`,
+which you can use to update the view. 
+
+Pass Environment Values
+```````````````````````
+From the LoginView, you can inject the ``.partitionvalue`` environment value
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:379]:
```suggestion
From the LoginView, you can inject the ``.partitionValue`` environment value 
```

Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:379]: It doesn't necessarily have to come from the LoginView. Partition values may also not always be based on the user id (though probably). So, it seems a bit specific to refer to "the LoginView" in this doc context.

**Review by cbush — COMMENTED** [review]
I find the environment-passing examples to be a bit more confusing than if they were perhaps not there. What about making one section or note that lets users know they can use [the environment(_:_:) view modifier](https://developer.apple.com/documentation/swiftui/view/environment(_:_:)) to pass the partitionValue down?

Commenting for now, will review again upon discussion.

File: source/sdk/ios/examples/sync-changes-between-devices.txt (hunk 9 above)
Comment (cbush) [reply source/sdk/ios/examples/sync-changes-between-devices.txt:407]: Scratch that. Self-addressed in the overall review comment.

File: source/sdk/ios/examples/sync-changes-between-devices.txt (hunk 7 above)
Comment (dacharyc) [reply source/sdk/ios/examples/sync-changes-between-devices.txt:350]: I used {+app+} here since we were talking about synced realms, but after thinking about this more, yeah, I think you're right and it should be general "Apps". Changing.

File: source/sdk/ios/examples/sync-changes-between-devices.txt [hunk 12]
```diff
@@ -328,6 +328,202 @@ users from opening a realm while offline.
...
+
+Open a Synced Realm with SwiftUI
+--------------------------------
+
+.. versionadded:: 10.12.0
+
+SwiftUI apps can use RealmSwift property wrappers to open {+realms+} and 
+populate views. These property wrappers provide a syntax that integrates 
+with SwiftUI to let developers declare whether their clients should download 
+changesets before opening {+realm+}s, or open {+realm+}s with data on the
```
Comment (cbush) [inline source/sdk/ios/examples/sync-changes-between-devices.txt:341]: (No comma when not changing the subject)
```suggestion
changesets before opening {+realm+}s or open {+realm+}s with data on the 
```

**Review by cbush — APPROVED** [review]
Ok, looks clearer now.

What if the examples came before the "pass environment values" sections? It still feels like it could be confusing if you don't know a more complete context is still coming.

And what if the "Pass environment values" section were shared between both auto-open and async open? (as a subheading of "Open ... with SwiftUI" I guess).

Up to you.
